### DISCUSSION TITLE
Custom CSS section

### DISCUSSION THREAD

**Discussion by christian-zeald** [outline/outline#2362] [Ideas]
Hi,  

Our company is currently using this to organize our documentations which is great so far.  Currently, I was tasked to modify some CSS styling to try and make it more customized to our standards.  I see that I can modify the `globals.js` to add the customizations to the style we require and just rebuild.  However, this would be overwritten whenever a new update is implemented unfortunately.  

A suggestion that comes into mind is to add a `custom.js` file that will always be included whenever a build/re-build is initiated and is always ignored by the repo.  In this file, we can add custom CSS.  I am unable to find a way to dynamically add this custom import without affecting the core files which will always be overwritten.

Please note that I am not an expert in node apps, I am just learning as I go.  Please do inform me if I missed out on a feature that enables me to do the customizations and will gladly utilize that instead.

Thanks for your time.  Again, great work on this app.  Cheers!

**Comment by tommoor** [comment]
Honestly, it's not really in-keeping with the philosophy of Outline to allow injection of arbitrary css/scripts, it's also difficult to achieve when hosting within docker (the recommended way of running Outline). 

I'd be interested to know what kind of style changes you're trying to make, we could certainly expose more structured theming options within the team admin.

**Reply by christian-zeald** [reply]
Hi Tommor,

Thanks so much for this information.  We shall just be adding the customizations manually before a re-build is done in an upgrade.  We were just wanting to add some additional space in between paragraphs and list items, just very minor ones.  

Keep up the great work!  Cheers!

**Reply by brentkelly** [reply]
Hi @tommoor - thanks heaps for your reply - much appreciated. Great wiki BTW - really tidy & 99% works well for us.

The issue we are trying to address here is the outline seems to strip all margin / whitespace from paragraphs, pre, lists etc. This results in documents that are either devoid of paragraphing & whitespace, or require double line breaks to achieve paragraphing.

I'm not a huge fan of this approach:
- I find it leads to inconsistent use of paragraphing/whitespace in documents as it relies on each document editor to use the appropriate amount of line breaks to manually space paragraphs out
- Double line breaks are not the equivalent of a normal paragraph break - they bigger - so you end up with documents with excessive white space between paragraphs.
- Inevitably it gets left off lists as most people don't think to use shift+enter to get a line break instead of enter which generates a new bullet.
- It's not portable - we can't use our outline wiki docs anywhere else by copy/pasting without reworking them.

An additional problem we have is we've imported a large amount of documentation from our past wiki, which uses normal paragraphing. So we either have to rework every document, or have a large amount of historic documentation with no paragraphing.

**Example of the issue**
Default styling:
![image](https://user-images.githubusercontent.com/2694025/127242068-76aeab09-c1b9-4eff-87d1-42f11f21890c.png)

vs what I'd expect (e.g. with Google docs, Word, general HTML, even github markdown):
![image](https://user-images.githubusercontent.com/2694025/127242218-350898e1-27d0-46fb-8340-2e3c4c0e9a39.png)

Here's the CSS overrides I've suggested:

```css
p {
    margin-bottom: 0.8rem !important;
}
li p {
    margin-bottom: 0.6rem !important;
}
pre {
    margin: 1rem 0 !important;
}
```

**Reply by lpar** [reply]
We're looking at Outline, and I would really like to see paragraph breaks as per the above comments. Everything crammed together is really hard to read.

**Reply by tommoor** [reply]
Are you also importing existing content?￼

**Reply by lpar** [reply]
Most of our existing content is in Markdown, and I also plan to carry on writing content in Markdown in Vim and then pasting it in.

Outline generates the paragraphs as \<p\> elements in the HTML just fine, it's just that the CSS says not to have spaces between paragraphs.

**Reply by tommoor** [reply]
That sounds like a yes, appreciate the context. I totally see the issue, unfortunately it's not as simple as just changing the css – we have millions of docs already created that assume the current spacing where it is under the users control. 

It could be that tweaking the way pasted/imported content is parsed to ensure the paragraph spacing would be a good solution.

**Reply by lpar** [reply]
The current paragraph spacing isn't under the user's control, though, unless I'm missing something -- there's no way for me to set that I want paragraph spacing to be non-zero. The only way to get a space between two pieces of text I can find is to type an entire empty paragraph between them, which is converted to a linebreak inside an empty paragraph (`<p><br></p>`). The paragraph still has no spacing, but the linebreak gives you some space on the page.

On the assumption that that is what everyone has been doing, I'd think it would be relatively easy to build migration code to remove the empty paragraph/linebreak combos once paragraphs have spacing "built in". They already seem to be treated differently, as they have a blank `class` attribute, unlike the normal paragraphs.

Now that I've poked at the editor a bit more, I notice that linebreaks in imported Markdown get stripped out, and there's no way to type a linebreak in the editor, so I guess those issues would also need fixing.

Ultimately, the support for Markdown is a massive part of the appeal of Outline, at least to me. It's so close to perfect that the remaining glitches are frustrating.

**Reply by brentkelly** [reply]
We will continue to just apply our own CSS overrides after each update as I find the 'double line break' approach nearly always creates messy, inconsistent spacing.

But yeah I totally get the predicament @tommoor. A rock and a hard place to fix that in a backwards compatible way.

The only options I see for fixing it in a way that doesn't mess up all the existing docs:
1. add an option for it - not ideal / messy.
2. force standard spacing and use some kind of upgrade script to port all existing docs to remove double line breaks - bags not me - can't see how its ever going to be thorough enough to guarantee not making a mess of existing docs.
3. make it easier for people to configure CSS / styling overrides that don't break every upgrade

1 & 2 aren't worth considering IMO. However #3 possibly could be worth a thought. Could be as simple as an `@import` in the SCSS.

Either way I'm not too worried - we've just documented the upgrade process for outline to include reapplying the CSS hacks which is working ok.

**Reply by dmajorhwl** [reply]
We've just started using Outline (paying for the cloud option) and this is something I instantly noticed. I'd appreciate an option to at least allow this to be a choice or something we can add as above as @brentkelly says, #3 would work well, but I don't know if that would open a can of worms for those that aren't familiar with CSS.

**Reply by lpar** [reply]
I wrote a TamperMonkey script to make it easy to test if removing `<p><br></p>` and putting spacing on paragraphs would work on random pages:

```js
// ==UserScript==
// @name         Outline fix proof of concept
// @namespace    http://tampermonkey.net/
// @version      0.1
// @description  Demonstrate fixing Outline documents to have paragraph spacing
// @author       mathew
// @match        https://*.getoutline.com/doc/*
// @icon         https://www.google.com/s2/favicons?domain=getoutline.com
// @grant        none
// ==/UserScript==

(function() {
    'use strict';
    console.log("Running tampermonkey script");
    const paras = document.evaluate("//p[not(text())]", document, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE);
    let p,i=0;
    while (p = paras.snapshotItem(i++)) {
        p.style.display = "none";
    }
    document.styleSheets[0].insertRule("p { padding: 0.5em 0; }");
})();
```

At least in my limited tests, it seems to improve the look of documents spaced using empty-paragraph-linefeeds. (I haven't yet worked out how to run it automatically after content load.)

**Reply by lpar** [reply]
If it helps, we're also interested in paid cloud hosting, so patching the CSS ourselves isn't a great answer.

**Reply by larslevie** [reply]
> That sounds like a yes, appreciate the context. I totally see the issue, unfortunately it's not as simple as just changing the css – we have millions of docs already created that assume the current spacing where it is under the users control.
> 
> It could be that tweaking the way pasted/imported content is parsed to ensure the paragraph spacing would be a good solution.

The lack of spacing between paragraphs is one of my biggest issues with Outline. In fact that is *exactly* the issue I came to solve when I found this discussion. I realize it sounds silly 😄 — but readability is serious, and manually adding the spacing with an empty paragraph doesn't seem like the correct solution, especially in the event of having to export content.

Even though it feels like a kludge from a product perspective, maybe you could support an option that tweaks the styles to add the spaces in. This would simply add spacing after paragraphs, lists, etc.

Update: we are on the hosted plan and so the hack-the-global-styles option isn't available to us.

**Reply by ReessKennedy** [reply]
Update here on a solution that doesn't require ongoing patching or mucking about with source code with each release: https://github.com/outline/outline/discussions/4206

Basically: 
1. Use Cloudflare
2. Install one of their "add css" apps and add your customizations
3. Custom styles get installed on your domain at the CDN level so they are global for all users!
4. Only maintenance is an update if selectors ever change on update. 

I deployed this successfully for my team for a number of mods to our install. 

Separately, for style mods only I like/want or am testing I use this cool little Chrome extension: 
https://chrome.google.com/webstore/detail/user-javascript-and-css/nbhcbdghjpllgmfilhnhkllmkecfmpld?hl=en

**Reply by winkelement** [reply]
> We will continue to just apply our own CSS overrides after each update as I find the 'double line break' approach nearly always creates messy, inconsistent spacing.
> 
> Either way I'm not too worried - we've just documented the upgrade process for outline to include reapplying the CSS hacks which is working ok.

@brentkelly Can you elaborate on how/where you modify the CSS? That would be super helpful as i am struggling to find the place to start with...

**Comment by ReessKennedy** [comment]
+1 voice support for custom CSS injection somewhere inside settings, for admins--instead of needing to hack an actual build. 

We'd like to just alter the link style and make them traditional blue links instead of black. 

Great work on this!

**Comment by lpar** [comment]
Just noting that it appears the built-in CSS was updated to include paragraph spacing, so this is now less urgent, at least from my point of view.

**Comment by EduNetEurope** [comment]
Here are more votes for CSS customisation:
https://github.com/outline/outline/discussions/4206

**Comment by nabeelio** [comment]
+1 on this

I want to modify the spacing, especially from headers to paragraphs - it's so cramped, no room to breath
